Add I2C HAL API overhaul design document

scartmell-arm opened the pull request:
## Description

This PR adds the Request For Comment document for the I2C HAL API overhaul.

This document is meant to serve as the design document for defining what we plan for the API.

Document readable [here](https://github.com/scartmell-arm/mbed-os/blob/rfc-i2c-overhaul/hal/rfcs/0001-i2c-overhaul.md).

Pull request type:
```
[ ] Fix
[ ] Refactor
[ ] New target
[x] Feature
[ ] Breaking change
```

@ithinuel  @c1728p9

Commit:
docs(hal-i2c): Add initial design document for the I2C HAL API overhaul

Commit:
docs(hal-i2c): Add modifications for timeout and multimaster support

Commit:
docs(hal-i2c): Update the RFC to reflect removing the DMAUsage parameter

Commit:
docs(hal-i2c): Add missing stop and last parameters to i2c_read/write

Commit:
docs(hal-i2c): Move address parameter into a seperate i2c_slave function

Commit:
docs(hal-i2c): Add additional defined function behaviours

ghost commented:
@AnotherButler

cmonr commented:
@mbed-os-core @ARMmbed/mbed-os-hal @c1728p9 @ithinuel @AnotherButler 

It would be good to get eyes on this PR since it's been open for a while.

AnotherButler commented:
I'll wait to edit this until some reviews are in and we know the content won't change too much.

Commit:
docs(hal-i2c): Make additional amendments to I2C API specification

- Add return value to i2c_frequency indicating configured frequency.
- Change return type of i2c_write/i2c_read from int to int32_t
- Add type suffix to i2c_slave_status enum.
- Rename i2c_slave_receive to i2c_slave_status.

Commit:
docs(hal-i2c): Add additional notes to synchronous read/write

Commit:
docs(hal-i2c): Add clock stretching field to target capabilities

Commit:
docs(hal-i2c): Add context pointer to async transfer function

ithinuel commented on hal/rfcs/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,484 @@
+ *           allow a user adjust its strategy in case the target cannot be
+ *           reached.
+ */
+void i2c_frequency(i2c_t *obj, uint32_t frequency);
```
I would also recommend to return the actual frequency used so that user can decide if the actual frequency used is acceptable for his use case.

ithinuel reviewed (changes requested):
Apart from the commentary about `i2c_frequency`, it looks good to me :)

Commit:
docs(hal-i2c): Fix return type of i2c_frequency function

0xc0170 commented:
@donatieng do you want to review or anyone else should or happy with the approvals as they are and we can progress with this design document?

cmonr commented:
@AnotherButler It seems that this PR has settled a bit.

SenRamakri commented:
@scartmell-arm - Whys is this document under hal, please move this to docs/design-documents/hal, you may create another folder there if you like. This document is not in the prescribed format all other design docs are. Please take a look at - https://github.com/ARMmbed/mbed-os/tree/master/docs/design-documents

Commit:
docs(hal-i2c): Move design document to the updated folder

ghost commented:
> @scartmell-arm - Whys is this document under hal, please move this to docs/design-documents/hal, you may create another folder there if you like. This document is not in the prescribed format all other design docs are. Please take a look at - https://github.com/ARMmbed/mbed-os/tree/master/docs/design-documents

Because it was initially written before either of those were added.

tumukei commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,484 @@
+
+- The implementation of the `i2c_byte_write` differs between platforms.
+
+  The API documentation indicates that the return value should be 0 on success, 1 on NAK, 2 on timeout. Not all platforms return a value or handle a timeout. Some platforms return a completely different set of status, the Atmel implementation returns an `I2C_ERROR ` which does not map to the documented return value.
```
The return value mixup also show up on the I2C class, as byte write and multiple byte write return different status values on ack/nack.

donatieng commented:
Hey @scartmell-arm, looks good but seems that this has diverged a bit from the implementation

jamesbeyond commented on docs/design-documents/hal/0001-i2c-overhaul.md:
```diff
@@ -0,0 +1,484 @@
+ *  When in master mode the operation consists of:
+ *   - Address the slave as a Master receiver.
+ *   - Receive data from the addressed slave.
+ *   - Generate a STOP condition if the specified `stop` field is true.
```
The read function only has a `last` field. maybe we should change it to `stop`, to make it consist with the write function

cmonr commented:
@scartmell-arm Any updates?

ghost commented:
> @scartmell-arm Any updates?

There were a number of minor changes made to the API after this was written, I'm in the process of updating it to match.

Commit:
docs(hal-i2c): Update design document to match recent API alterations

cmonr commented:
@scartmell-arm Are there more updates in the pipeline, or should we start poking reviewers again?

SenRamakri reviewed (changes requested):
@scartmell-arm - Please follow the design-document template if this needs to be put under docs/design-documents directory - https://github.com/ARMmbed/mbed-os/blob/master/docs/design-documents/design_template.md

Commit:
docs(hal-i2c): Update the comments of the read/write `STOP` parameter
